vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! Common code that's shared across all migration code.
+//! Each network upgrade / state migration code lives in their own module.
+
+use actor_interface::{actorv3, actorv4};
+use address::Address;
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_blockstore::BlockStore;
+use state_tree::StateTree;
+use vm::{ActorState, TokenAmount};
+
+use async_std::sync::Arc;
+use rayon::ThreadPoolBuildError;
+use std::collections::{HashMap, HashSet};
+use std::default::Default;
+
+pub mod nv12;
+
+pub const ACTORS_COUNT: usize = 11;
+
+pub type Migrator<BS> = Arc<dyn ActorMigration<BS> + Send + Sync>;
+pub type MigrationResult<T> = Result<T, MigrationError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    // FIXME: use underlying concrete types when possible.
+    #[error("Failed creating job for state migration: {0}")]
+    MigrationJobCreate(String),
+    #[error("Failed running job for state migration: {0}")]
+    MigrationJobRun(String),
+    #[error("Flush failed post migration: {0}")]
+    FlushFailed(String),
+    #[error("Failed writing to blockstore: {0}")]
+    BlockStoreWrite(String),
+    #[error("Failed reading from blockstore: {0}")]
+    BlockStoreRead(String),
+    #[error("Migrator not found for cid: {0}")]
+    MigratorNotFound(Cid),
+    #[error("Failed updating new actor state: {0}")]
+    SetActorState(String),
+    #[error("State tree creation failed")]
+    StateTreeCreation(String),
+    #[error("Incomplete migration specification with {0} code CIDs")]
+    IncompleteMigrationSpec(usize),
+    #[error("Thread pool creation failed: {0}")]
+    ThreadPoolCreation(ThreadPoolBuildError),
+    #[error("Migration failed")]
+    Other,
+}
+
+pub struct StateMigration<BS> {
+    migrations: HashMap<Cid, Migrator<BS>>,
+    deferred_code_ids: HashSet<Cid>,
+}
+
+impl<BS: BlockStore + Send + Sync> Default for StateMigration<BS> {
+    fn default() -> Self {
```

Comment:
I wouldnt implement default on this.

Reply:
Yeah, I had `new` implemented but clippy asked me to do it. Can we disable that lint?

Reply:
Either that or we could just default to default. I don't mind default over new, even though it's a little more verbose sometimes.

Reply:
I guess it doesn't make sense to have `nil_migrator` with the `default` method. I'll keep only a `new` method. Also, migrations must be initialized in the `migrations` map specific to a given migration and not here (for this it can be in `nv12/mod.rs`). At present, we are initializing it with specific actor v3 -> v4, which doesn't make it a generic `new` method.